core/deployment/src/main/java/io/quarkus/deployment/recording/BytecodeRecorderImpl.java

```diff
@@ -1231,12 +1231,12 @@ ResultHandle createValue(MethodContext context, MethodCreator method, ResultHand
                         out = method.newInstance(ofConstructor(param.getClass()));
                     } catch (NoSuchMethodException e) {
                         //fallback for collection types, such as unmodifiableMap
-                        if (expectedType == Map.class) {
+                        if (Map.class.isAssignableFrom(expectedType)) {
                             out = method.newInstance(ofConstructor(LinkedHashMap.class));
-                        } else if (expectedType == List.class) {
+                        } else if (List.class.isAssignableFrom(expectedType)) {
                             out = method.newInstance(ofConstructor(ArrayList.class));
-                        } else if (expectedType == Set.class) {
-                            out = method.newInstance(ofConstructor(Set.class));
+                        } else if (Set.class.isAssignableFrom(expectedType)) {
+                            out = method.newInstance(ofConstructor(HashSet.class));
```

Comment:
If we went for a `LinkedHashMap` for maps, maybe we should go for a `LinkedHashSet` for sets.

Reply:
Ideally we should probably check for SortedMap and SortedSet as well and use Tree Map/Set, if we want to be as correct as possible.

Reply:
Done